File: Lexer.hs
{-# OPTIONS_GHC -Wall -Wextra #-}
module Main where

import Data.Char (isAsciiLower, isDigit)

data TokenName =
    Number
  | Operator
  | Id
  | Lparen
  | Rparen
  | Comma
  deriving (Show)

type Token = (TokenName, String)
type State = Int

stateToToken :: String -> State -> Maybe Token
stateToToken inp state =
  case state of
    1  -> Just (Number  , inp)
    2  -> Just (Number  , inp)
    5  -> Just (Number  , inp)
    6  -> Just (Operator, inp)
    7  -> Just (Id      , inp)
    8  -> Just (Lparen  , inp)
    9  -> Just (Rparen  , inp)
    10 -> Just (Comma   , inp)
    _  -> Nothing

transitionTable :: State -> Char -> Maybe State
transitionTable state char =
  case state of
    0 | isLetter char       -> Just 7
      | char == '('         -> Just 8
      | char == ')'         -> Just 9
      | char == ','         -> Just 10
      | char `elem` "+*/^%-"-> Just 6
      | isDigit char        -> Just 1
    1 | isDigit char        -> Just 1
      | char == '.'         -> Just 2
    2 | isDigit char        -> Just 2
      | char `elem` "eE"    -> Just 3
    3 | char `elem` "+-"    -> Just 4
      | isDigit char        -> Just 5
    4 | isDigit char        -> Just 5
    5 | isDigit char        -> Just 5
    7 | isDigit char || isLetter char
                            -> Just 7
    _                       -> Nothing
  where
  isLetter c = isAsciiLower c || c == '_'

data DFAAction = Continue | Restart

type StateWIO = (String, String, State)
              -- input , output, st
type DFAState = (StateWIO, Maybe StateWIO)
              -- (input, st)    , last accepting state

dfaStep :: DFAState -> (DFAAction, DFAState)
dfaStep ds =
  let ((input, output, st), lac) = ds
      nextState = transitionTable st (head input)
      newAcceptingState =
        maybe lac (const $ Just (input, output, st))
        $ stateToToken output st
      input' = tail input
      output' = output <> [head input]
  in if null input
     then (Restart, (undefined, newAcceptingState))
     else  maybe (Restart, (undefined, newAcceptingState))
                 (\x -> (Continue, ((input', output', x), newAcceptingState)))
                 nextState

dfaLoop :: DFAState -> [Maybe Token]
dfaLoop st =
  case dfaStep st of
    (Continue, st') -> dfaLoop st'
    (Restart, (_, Just accSt))
        -> stateToToken output st' : dfaLoop ((input, "", 0), Nothing)
        where (input, output, st') = accSt
    (Restart, (_, Nothing)) -> []

mainLoop :: String -> String
mainLoop input = show $ dfaLoop ((input, "", 0), Nothing)

main :: IO ()
main = interact mainLoop

File: lexer.cpp
#include "lexer.h"
#include <stdexcept>

int transTable(char c, int st) {
  switch (st) {
  case 0:
    if (c >= '0' && c <= '9')
      return 1;
    else if (c == '+' || c == '*' || c == '/' || c == '-' || c == '^')
      return 6;
    else if ((c >= 'a' && c <= 'z') || c == '_')
      return 7;
    else if (c == '(')
      return 8;
    else if (c == ')')
      return 9;
    else if (c == ',')
      return 10;
    else if (c == ' ' || c == '\t' || c == '\n' || c == '\r')
      return 0;
    break;
  case 1:
    if (c >= '0' && c <= '9')
      return 1;
    else if (c == '.')
      return 2;
    else if (c == 'e' || c == 'E')
      return 3;
    break;
  case 2:
    if (c >= '0' && c <= '9')
      return 2;
    else if (c == 'e' || c == 'E')
      return 3;
    break;
  case 3:
    if (c >= '0' && c <= '9')
      return 5;
    else if (c == '+' || c == '-')
      return 4;
    break;
  case 4:
    if (c >= '0' && c <= '9')
      return 5;
    break;
  case 5:
    if (c >= '0' && c <= '9')
      return 5;
    break;
  case 6:
    break;
  case 7:
    if ((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_')
      return 7;
    break;
  }
  return -1;
}

Lexer::Lexer(const std::string &input) : input(input), curChIx(0) {}

Token Lexer::getNextToken() {
  std::string buf;
  std::size_t lastAccChIx = curChIx;
  int curSt = 0;
  Token resultToken{TokenType::Invalid};
  while (curSt != -1) {
    if (curSt == 1 || curSt == 2 || curSt == 5) {
      resultToken.type = TokenType::Number;
      resultToken.attribute = buf;
      lastAccChIx = curChIx;
    } else if (curSt == 6) {
      resultToken.type = TokenType::Operator;
      resultToken.attribute = buf;
      lastAccChIx = curChIx;
    } else if (curSt == 7) {
      resultToken.type = TokenType::Id;
      resultToken.attribute = buf;
      lastAccChIx = curChIx;
    } else if (curSt == 8) {
      resultToken.type = TokenType::LParen;
      resultToken.attribute = buf;
      lastAccChIx = curChIx;
    } else if (curSt == 9) {
      resultToken.type = TokenType::RParen;
      resultToken.attribute = buf;
      lastAccChIx = curChIx;
    } else if (curSt == 10) {
      resultToken.type = TokenType::Comma;
      resultToken.attribute = buf;
      lastAccChIx = curChIx;
    }
    if (curChIx >= input.size())
      break;
    char curCh = input[curChIx];
    curChIx += 1;
    curSt = transTable(curCh, curSt);
    if (curSt > 0) // пропуск пробельных символов
      buf += curCh;
  }
  // if (resultToken.type == TokenType::Invalid)
  //   throw new std::runtime_error("Invalid token in input " + input);
  curChIx = lastAccChIx;
  return resultToken;
}
